File: baekjoon/gold/baekjoon_2589.py
# 보물섬
# 상하좌우 육지로만 이동 가능, 한 칸 이동하는데 +1
# 보물은 가장 먼 거리에 묻혀있고, 최대 거리를 구해라
# 완전탐색?
# 좌표 하나를 선택해서 최대거리를 구하고 answer 저장
# 다른 좌표를 하나 선택해서 최대거리를 구하고 answer 저장
# 모든 L 좌표에서 최대거리를 구하고 가장 많이 걸리는 시간을 답으로 도출


from collections import deque
import sys

n,m = map(int,sys.stdin.readline().split())
graph = [list(map(str, sys.stdin.readline())) for _ in range(n)]

def bfs(a,b):
    dx = [-1,0,1,0]
    dy = [0,1,0,-1]
    visited = [[False]*m for _ in range(n)]
    visited[a][b] = True
    dist = [[0]*m for _ in range(n)]

    q = deque()
    q.append((a,b))

    global answer

    while q:
        y,x = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0<=nx<m and 0<=ny<n:
                if not visited[ny][nx] and graph[ny][nx] == 'L':
                    visited[ny][nx] = True
                    dist[ny][nx] = dist[y][x] + 1
                    q.append((ny,nx))

    cnt = max(map(max, dist))
    answer = max(cnt, answer)
        
answer = 0

for i in range(n):
    for j in range(m):
        if graph[i][j] == 'L':  
            bfs(i,j)

print(answer)

File: baekjoon/gold/baekjoon_2589.java
import java.io.*;
import java.util.*;

public class baekjoon_2589{

    static int n,m;
    static int answer, cnt;
    static char[][] graph;
    static int[][] dist;
    static boolean[][] visited;
    static int[] dx = {-1,0,1,0};
    static int[] dy = {0,1,0,-1};
    static ArrayList<Integer> result = new ArrayList<Integer>();

    public static void main(String[] args) throws IOException{
            BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
            String[] str = bf.readLine().split(" ");

            n = Integer.parseInt(str[0]);
            m = Integer.parseInt(str[1]);

            graph = new char[n][m];
            visited = new boolean[n][m];
            dist = new int[n][m];

            answer = 0;

            for(int i=0; i<n; i++){
                char[] ch = bf.readLine().toCharArray();
                for (int j=0; j<m; j++){
                    graph[i][j] = ch[j];
                }
            }

            for(int i=0;i<n;i++){
                for(int j=0;j<m;j++){
                    if (graph[i][j] == 'L'){
                        bfs(i,j);
                    }
                    for(int k=0; k<n;k++){
                        Arrays.fill(visited[k],false);
                        Arrays.fill(dist[k],0);
                    }
                    
                }
            }

            System.out.println(Collections.max(result));
        
    }

    public static void bfs(int a,int b){

        cnt = 0;

        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{a,b});
        visited[a][b] = true;

        while(!q.isEmpty()){
            int x = q.poll()[0];
            int y = q.poll()[1];
            
            for(int i=0; i<4;i++){
                int nx = x + dx[i];
                int ny = y + dy[i];

                if(0<=nx && 0<=ny && nx<m && ny<n){
                    if(!visited[nx][ny] && graph[nx][ny] == 'L'){
                        dist[nx][ny] = dist[x][y] + 1;
                        visited[nx][ny] = true;
                        q.add(new int[] {nx,ny});

                    }
                }
            }
        }
        

    }
    
}
